src/core/server/saved_objects/saved_objects_service.ts

```diff
@@ -0,0 +1,116 @@
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { CoreService } from 'src/core/types';
+import { first } from 'rxjs/operators';
+import { KibanaMigrator } from './migrations';
+import { CoreContext } from '../core_context';
+import { LegacyServiceSetup } from '../legacy/legacy_service';
+import { ElasticsearchServiceSetup } from '../elasticsearch';
+import { KibanaConfig } from '../kibana_config';
+import { retryCallCluster } from '../elasticsearch/retry_call_cluster';
+import { Logger } from '..';
+
+/**
+ * @public
+ */
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface SavedObjectsServiceSetup {}
+
+/**
+ * @public
+ */
+export interface SavedObjectsServiceStart {
+  migrator: Pick<KibanaMigrator, keyof KibanaMigrator>;
+}
+
+/** @internal */
+export interface SavedObjectsSetupDeps {
+  legacy: LegacyServiceSetup;
+  elasticsearch: ElasticsearchServiceSetup;
+}
+
+/** @internal */
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface SavedObjectsStartDeps {}
+
+export class SavedObjectsService
+  implements CoreService<SavedObjectsServiceSetup, SavedObjectsServiceStart> {
+  private migrator: KibanaMigrator | undefined;
+  logger: Logger;
+
+  constructor(private readonly coreContext: CoreContext) {
+    this.logger = coreContext.logger.get('savedobjects-service');
+  }
+
+  public async setup(coreSetup: SavedObjectsSetupDeps) {
+    this.logger.debug('Setting up SavedObjects service');
+
+    const {
+      savedObjectSchemas,
+      savedObjectMappings,
+      savedObjectMigrations,
+      savedObjectValidations,
+    } = await coreSetup.legacy.uiExports;
+
+    const adminClient = await coreSetup.elasticsearch.adminClient$.pipe(first()).toPromise();
+
+    const kibanaConfig = await this.coreContext.configService
+      .atPath<KibanaConfig>('kibana')
+      .pipe(first())
+      .toPromise();
+
+    this.migrator = new KibanaMigrator({
+      savedObjectSchemas,
+      savedObjectMappings,
+      savedObjectMigrations,
+      savedObjectValidations,
+      logger: this.coreContext.logger.get('migrations'),
+      kibanaVersion: this.coreContext.env.packageInfo.version,
+      config: coreSetup.legacy.pluginExtendedConfig,
+      kibanaConfig: (kibanaConfig as any) as KibanaConfig,
+      callCluster: retryCallCluster(adminClient.callAsInternalUser.bind(adminClient)),
+    });
+
+    return ({} as any) as Promise<SavedObjectsServiceSetup>;
```

Comment:
why we need this workaround? lifecycles are expected to return void

Reply:
It's because I specified an empty `SavedObjectsServiceSetup` interface. I could remove it, but I'm using this in my follow-up PR to expose the saved objects client.